crates/astria-sequencer/src/app.rs

```diff
@@ -0,0 +1,371 @@
+use std::sync::Arc;
+
+use anyhow::Result;
+use penumbra_storage::{
+    ArcStateDeltaExt,
+    Snapshot,
+    StateDelta,
+    Storage,
+};
+use tendermint::abci::{
+    self,
+    Event,
+};
+use tracing::{
+    info,
+    instrument,
+};
+
+use crate::{
+    accounts::{
+        component::AccountsComponent,
+        types::{
+            Address,
+            Balance,
+        },
+    },
+    component::Component,
+    state_ext::{
+        StateReadExt as _,
+        StateWriteExt as _,
+    },
+    transaction::{
+        ActionHandler as _,
+        Transaction,
+    },
+};
+
+/// The application hash, used to verify the application state.
+/// TODO: this may not be the same as the state root hash?
+pub type AppHash = penumbra_storage::RootHash;
+
+/// The inter-block state being written to by the application.
+type InterBlockState = Arc<StateDelta<Snapshot>>;
+
+/// The genesis state for the application.
+#[derive(Debug, serde::Deserialize, Default)]
+pub struct GenesisState {
+    pub accounts: Vec<(Address, Balance)>,
+}
+
+/// The Sequencer application, written as a bundle of [`Component`]s.
+///
+/// Note: this is called `App` because this is a Tendermint ABCI application,
+/// and implements the state transition logic of the chain.
+///
+/// See also https://github.com/penumbra-zone/penumbra/blob/9cc2c644e05c61d21fdc7b507b96016ba6b9a935/app/src/app/mod.rs#L42.
+#[derive(Clone, Debug)]
+pub struct App {
+    state: InterBlockState,
+}
+
+impl App {
+    pub fn new(snapshot: Snapshot) -> Self {
+        tracing::debug!("initializing App instance");
+
+        // We perform the `Arc` wrapping of `State` here to ensure
+        // there should be no unexpected copies elsewhere.
+        let state = Arc::new(StateDelta::new(snapshot));
+
+        Self {
+            state,
+        }
+    }
+
+    #[instrument(name = "App:init_chain", skip(self))]
+    pub async fn init_chain(&mut self, genesis_state: GenesisState) -> Result<()> {
+        // allocate to hard-coded accounts for testing
+        let mut accounts = genesis_state.accounts;
+        accounts.append(&mut default_genesis_accounts());
+        let genesis_state = GenesisState {
+            accounts,
+        };
+
+        let mut state_tx = self
+            .state
+            .try_begin_transaction()
+            .expect("state Arc should not be referenced elsewhere");
+
+        state_tx.put_block_height(0);
+
+        // call init_chain on all components
+        AccountsComponent::init_chain(&mut state_tx, &genesis_state).await?;
```

Comment:
Should we have some sort of Component manager to define modules and the order they will be called in similar to the Cosmos [manager](https://docs.cosmos.network/v0.47/building-modules/module-manager#manager)?

Reply:
the cosmos manager seems more needed because it's an SDK, and it needs to figure out what modules exist at runtime, whereas for us it's fine to just write them in and do it at compile time (this is what penumbra does too)